### PR TITLE
Update `TextCatBOW` to use the fixed `SparseLinear` layer

### PR DESCRIPTION
**Pull request by danieldk**
## Description

A while ago, we fixed the `SparseLinear` layer to use all available
parameters: https://github.com/explosion/thinc/pull/754

This change updates `TextCatBOW` to `v3` which uses the new
`SparseLinear_v2` layer. This results in a sizeable improvement on a
text categorization task that was tested.

While at it, this `spacy.TextCatBOW.v3` also adds the `length_exponent`
option to make it possible to change the hidden size. Ideally, we'd just
have an option called `length`. But the way that `TextCatBOW` uses
hashes results in a non-uniform distribution of parameters when the
length is not a power of two.

### Types of change

Bugfix

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/tests/pipeline/test_textcat.py [hunk 1]
```diff
@@ -451,11 +451,16 @@ def test_no_resize(name, textcat_config):
...
+        # BOW V2
         ("textcat", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": True, "no_output_layer": False, "ngram_size": 3}),
         ("textcat", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": True, "no_output_layer": True, "ngram_size": 3}),
         ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": False, "no_output_layer": False, "ngram_size": 3}),
         ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": False, "no_output_layer": True, "ngram_size": 3}),
+        # BOW V3
+        ("textcat", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": True, "no_output_layer": False, "ngram_size": 3}),
+        ("textcat", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": True, "no_output_layer": True, "ngram_size": 3}),
+        ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": False, "no_output_layer": False, "ngram_size": 3}),
+        ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": False, "no_output_layer": True, "ngram_size": 3}),
```
Comment (svlandeg) [inline spacy/tests/pipeline/test_textcat.py:463]: Let's not copy all these tests, it'll blow up the test suite if we always do. We've also got `test_textcat_configs` that specifically tests older architectures. Here, I suggest to just update to the new one.

File: spacy/tests/pipeline/test_textcat.py [hunk 2]
```diff
@@ -480,11 +485,16 @@ def test_resize(name, textcat_config):
...
+        # BOW v2
         ("textcat", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": True, "no_output_layer": False, "ngram_size": 3}),
         ("textcat", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": True, "no_output_layer": True, "ngram_size": 3}),
         ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": False, "no_output_layer": False, "ngram_size": 3}),
         ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v2", "exclusive_classes": False, "no_output_layer": True, "ngram_size": 3}),
+        # BOW v3
+        ("textcat", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": True, "no_output_layer": False, "ngram_size": 3}),
+        ("textcat", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": True, "no_output_layer": True, "ngram_size": 3}),
+        ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": False, "no_output_layer": False, "ngram_size": 3}),
+        ("textcat_multilabel", {"@architectures": "spacy.TextCatBOW.v3", "exclusive_classes": False, "no_output_layer": True, "ngram_size": 3}),
```
Comment (svlandeg) [inline spacy/tests/pipeline/test_textcat.py:497]: Here as well I'd just update to use v3.

**Review by svlandeg — COMMENTED** [review]
Looks great!

> Ideally, we'd just have an option called `length`. But the way that TextCatBOW uses
hashes results in a non-uniform distribution of parameters when the
length is not a power of two.

One alternative would be to take the highest power of two fitting within `length` if we don't want to expose these internals too much to users.

**Comment by adrianeboyd** [comment]
If you do keep `length_exponent` as a direct setting, I would vote strongly for a different name that's easier for users to understand. I don't immediately know what, though...

**Comment by danieldk** [comment]
> One alternative would be to take the highest power of two fitting within `length` if we don't want to expose these internals too much to users.

Rounding up seems like a good strategy. I'll update the PR to do that.

File: spacy/ml/models/textcat.py [hunk 3]
```diff
@@ -95,10 +97,48 @@ def build_bow_text_classifier(
...
+    exclusive_classes: bool,
+    ngram_size: int,
+    no_output_layer: bool,
+    length: int = 262144,
+    nO: Optional[int] = None,
+) -> Model[List[Doc], Floats2d]:
+    if length < 1:
+        raise ValueError(Errors.E1056.format(length=length))
+
+    # Find k such that 2**(k-1) < length <= 2**k.
```
Comment (adrianeboyd) [inline spacy/ml/models/textcat.py:121]: I'd include a note about why you are doing this (and I think it's confusing that you're reusing the variable name).

File: spacy/ml/models/textcat.py [hunk 4]
```diff
@@ -95,10 +97,48 @@ def build_bow_text_classifier(
...
+        sparse_linear=SparseLinear(nO=nO),
+    )
+
+
+@registry.architectures("spacy.TextCatBOW.v3")
+def build_bow_text_classifier_v3(
+    exclusive_classes: bool,
+    ngram_size: int,
+    no_output_layer: bool,
+    length: int = 262144,
```
Comment (adrianeboyd) [inline spacy/ml/models/textcat.py:115]: To line up more with `SparseLinear`:

```suggestion
    length: int = 2 ** 18,
```

**Review by adrianeboyd — COMMENTED** [review]
I feel like the power-of-2 bit could be emphasized a bit more for people coming back to this code in the future.

**Pull request merged**